File: VB/MyGridCtrl/MyBandedGridViewInfo.vb
Imports Microsoft.VisualBasic
Imports DevExpress.XtraGrid
Imports DevExpress.XtraGrid.Drawing
Imports DevExpress.XtraGrid.Views.BandedGrid
Imports DevExpress.XtraGrid.Views.BandedGrid.ViewInfo
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.Linq

Namespace DevExpress.S170736
	Public Class MyBandedGridViewInfo
		Inherits BandedGridViewInfo
		Public Sub New(ByVal gridView As BandedGridView)
			MyBase.New(gridView)
		End Sub
		Protected Overrides Function CalcGroupFooterHeight() As Integer
			Dim myView = CType(View, MyBandedGridView)
			If myView.GetGroupFooterRowCount() <= 0 Then
				Return 0
			End If
			Return MyBase.CalcGroupFooterHeight() * myView.GetGroupFooterRowCount()
		End Function
		Protected Overrides Function GetMaxColumnFooterCount() As Integer
			Return (CType(View, MyBandedGridView)).GetGroupFooterRowCount()
		End Function
		Protected Overrides Sub CalcRowCellsFooterInfo(ByVal fi As GridRowFooterInfo, ByVal ri As GridRowInfo)
			AddColumnsInfo(fi.RowHandle)
			MyBase.CalcRowCellsFooterInfo(fi, ri)
			ChangeDisplayText(fi)
			RemoveColumnsInfo()
		End Sub
		Private Sub AddColumnsInfo(ByVal rowHandle As Integer)
            rowHandle = If(rowHandle >= 0, -1, rowHandle)
            Dim groupSummary = View.GetGroupSummaryValues(rowHandle)
            Dim myGroupSummaryList = New List(Of DictionaryEntry)()
            For Each item As DictionaryEntry In groupSummary
                myGroupSummaryList.Add(item)
            Next item
            Dim query = From item In myGroupSummaryList
                Let summary = CType(item.Key, GridGroupSummaryItem)
                Group item By summary.ShowInGroupColumnFooter
                Into Group
            For Each group In query
                For i = 1 To group.Group.Count - 1
                    Dim sourceCInfo = ColumnsInfo(group.ShowInGroupColumnFooter)
                    If sourceCInfo Is Nothing Then
                        Continue For
                    End If
                    Dim cellInfo = New GridColumnInfoArgs(sourceCInfo.Column)
                    cellInfo.Assign(sourceCInfo)
                    cellInfo.Tag = TryCast(group.Group(i).Key, GridGroupSummaryItem)
                    cellInfo.Info.StartRow = i
                    ColumnsInfo.Add(cellInfo)
                Next i
            Next group
		End Sub
		Private Sub AddNewColumnsInfo(ByVal group As List(Of DictionaryEntry), ByVal i As Integer, ByVal initCellInfo As GridColumnInfoArgs)
			Dim cellInfo = New GridColumnInfoArgs(initCellInfo.Column)
			cellInfo.Assign(initCellInfo)
			cellInfo.Tag = TryCast(group(i).Key, GridGroupSummaryItem)
			cellInfo.Info.StartRow = i
			ColumnsInfo.Add(cellInfo)
		End Sub
		Private Sub ChangeDisplayText(ByVal fi As GridRowFooterInfo)
			For Each fci As GridFooterCellInfoArgs In fi.Cells
				Dim rowSummaryItem = View.GetRowSummaryItem(fi.RowHandle, fci.Column)
				If rowSummaryItem.Key Is Nothing Then
					Continue For
				End If
				Dim item = TryCast(fci.ColumnInfo.Tag, GridGroupSummaryItem)
				If item Is Nothing Then
					fci.Visible = If((CType(rowSummaryItem.Key, GridGroupSummaryItem)).SummaryType = Data.SummaryItemType.None, False, True)
					Continue For
				End If
				fci.Visible = If(item.SummaryType = Data.SummaryItemType.None, False, True)
				ChangeFooterCellDisplayText(fi, fci, item)
			Next fci
		End Sub
		Private Sub ChangeFooterCellDisplayText(ByVal fi As GridRowFooterInfo, ByVal fci As GridFooterCellInfoArgs, ByVal item As GridGroupSummaryItem)
			Dim dEntry = GetCustomSummaryItemValue(fi.RowHandle, fci.Column, item)
			fci.Visible = If(item.SummaryType = Data.SummaryItemType.None, False, True)
			fci.Value = dEntry.Value
			fci.DisplayText = (CType(dEntry.Key, GridSummaryItem)).GetDisplayText(dEntry.Value, False)
		End Sub
		Private Function GetCustomSummaryItemValue(ByVal rowHandle As Integer, ByVal column As DevExpress.XtraGrid.Columns.GridColumn, ByVal summaryItem As Object) As DictionaryEntry
			If column Is Nothing Then
				Return New DictionaryEntry()
			End If
			Dim summary = View.GetGroupSummaryValues(rowHandle)
			If summary Is Nothing Then
				Return New DictionaryEntry()
			End If
			For Each dEntry As DictionaryEntry In summary
				Dim item = TryCast(dEntry.Key, GridGroupSummaryItem)
				If item.ShowInGroupColumnFooter IsNot column OrElse item IsNot summaryItem Then
					Continue For
				End If
				Return dEntry
			Next dEntry
			Return New DictionaryEntry()
		End Function
		Private Sub RemoveColumnsInfo()
			Dim count = GetCount()
			If count < 0 Then
				Return
			End If
			Do While ColumnsInfo.Count > count
				ColumnsInfo.RemoveAt(count)
			Loop
		End Sub
		Private Function GetCount() As Integer
			For i = 0 To ColumnsInfo.Count - 1
				Dim item = TryCast(ColumnsInfo(i).Tag, GridGroupSummaryItem)
				If item Is Nothing Then
					Continue For
				End If
				Return i
			Next i
			Return -1
		End Function
		Protected Overrides Sub UpdateRowFooterInfo(ByVal ri As GridRowInfo, ByVal fi As GridRowFooterInfo)
			MyBase.UpdateRowFooterInfo(ri, fi)
			ChangeDisplayText(fi)
		End Sub
	End Class
End Namespace

File: CS/MyGridCtrl/MyBandedGridViewInfo.cs
using DevExpress.XtraGrid;
using DevExpress.XtraGrid.Drawing;
using DevExpress.XtraGrid.Views.BandedGrid;
using DevExpress.XtraGrid.Views.BandedGrid.ViewInfo;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;

namespace DevExpress.S170736 {
    public class MyBandedGridViewInfo : BandedGridViewInfo {
        public MyBandedGridViewInfo(BandedGridView gridView)
            : base(gridView) {
        }
        protected override int CalcGroupFooterHeight() {
            var myView = (MyBandedGridView)View;
            if (myView.GetGroupFooterRowCount() <= 0) {
                return 0;
            }
            return base.CalcGroupFooterHeight() * myView.GetGroupFooterRowCount();
        }
        protected override int GetMaxColumnFooterCount() {
            return ((MyBandedGridView)View).GetGroupFooterRowCount();
        }
        protected override void CalcRowCellsFooterInfo(GridRowFooterInfo fi, GridRowInfo ri) {
            AddColumnsInfo(fi.RowHandle);
            base.CalcRowCellsFooterInfo(fi, ri);
            ChangeDisplayText(fi);
            RemoveColumnsInfo();
        }
        private void AddColumnsInfo(int rowHandle) {
            rowHandle = rowHandle >= 0 ? -1 : rowHandle;
            var groupSummary = View.GetGroupSummaryValues(rowHandle);
            var myGroupSummaryList = new List<DictionaryEntry>();
            foreach (DictionaryEntry item in groupSummary) {
                myGroupSummaryList.Add(item);
            }
            var query = from item in myGroupSummaryList
                        let summary = (GridGroupSummaryItem)item.Key
                        group item by summary.ShowInGroupColumnFooter;
            foreach (var group in query) {
                var myGroup = group.ToList();
                for (var i = 1; i < myGroup.Count; i++) {
                    var sourceCInfo = ColumnsInfo[group.Key];
                    if (sourceCInfo == null) {
                        continue;
                    }
                    AddNewColumnsInfo(myGroup, i, sourceCInfo);
                }
            }            
        }
        private void AddNewColumnsInfo(List<DictionaryEntry> group, int i, GridColumnInfoArgs initCellInfo) {
            var cellInfo = new GridColumnInfoArgs(initCellInfo.Column);
            cellInfo.Assign(initCellInfo);
            cellInfo.Tag = group[i].Key as GridGroupSummaryItem;
            cellInfo.Info.StartRow = i;
            ColumnsInfo.Add(cellInfo);
        }
        private void ChangeDisplayText(GridRowFooterInfo fi) {
            foreach (GridFooterCellInfoArgs fci in fi.Cells) {
                var rowSummaryItem = View.GetRowSummaryItem(fi.RowHandle, fci.Column);
                if (rowSummaryItem.Key == null) {
                    continue;
                }
                var item = fci.ColumnInfo.Tag as GridGroupSummaryItem;
                if (item == null) {
                    fci.Visible = ((GridGroupSummaryItem)rowSummaryItem.Key).SummaryType == Data.SummaryItemType.None ? false : true;
                    continue;
                }
                fci.Visible = item.SummaryType == Data.SummaryItemType.None ? false : true;
                ChangeFooterCellDisplayText(fi, fci, item);
            }
        }
        private void ChangeFooterCellDisplayText(GridRowFooterInfo fi, GridFooterCellInfoArgs fci, GridGroupSummaryItem item) {
            var dEntry = GetCustomSummaryItemValue(fi.RowHandle, fci.Column, item);
            fci.Visible = item.SummaryType == Data.SummaryItemType.None ? false : true;
            fci.Value = dEntry.Value;
            fci.DisplayText = ((GridSummaryItem)dEntry.Key).GetDisplayText(dEntry.Value, false);
        }
        private DictionaryEntry GetCustomSummaryItemValue(int rowHandle, DevExpress.XtraGrid.Columns.GridColumn column, object summaryItem) {
            if (column == null) {
                return new DictionaryEntry();
            }
            var summary = View.GetGroupSummaryValues(rowHandle);
            if (summary == null) {
                return new DictionaryEntry();
            }
            foreach (DictionaryEntry dEntry in summary) {
                var item = dEntry.Key as GridGroupSummaryItem;
                if (item.ShowInGroupColumnFooter != column || item != summaryItem) {
                    continue;
                }
                return dEntry;
            }
            return new DictionaryEntry();
        }
        private void RemoveColumnsInfo() {
            var count = GetCount();
            if (count < 0) {
                return;
            }
            while (ColumnsInfo.Count > count) {
                ColumnsInfo.RemoveAt(count);
            }
        }
        private int GetCount() {
            for (var i = 0; i < ColumnsInfo.Count; i++) {
                var item = ColumnsInfo[i].Tag as GridGroupSummaryItem;
                if (item == null) {
                    continue;
                }
                return i;
            }
            return -1;
        }
        protected override void UpdateRowFooterInfo(GridRowInfo ri, GridRowFooterInfo fi) {
            base.UpdateRowFooterInfo(ri, fi);
            ChangeDisplayText(fi);
        }
    }
}
